File: Sorting/mergesort.rb
def mergesort(unsorted, low_index, high_index)
  if low_index < high_index
    mid = low_index + (high_index - low_index)/2
    mergesort(unsorted, low_index, mid)
    mergesort(unsorted, mid + 1, high_index)
    merge(unsorted, low_index, mid, high_index)
  end
end

def merge(unsorted, low_index, mid, high_index)
  original = unsorted.dup
  a = low_index
  b = mid + 1
  z = low_index
  while a <= mid and b <= high_index
    if original[a] <= original[b]
      unsorted[z] = original[a]
      a += 1
    else
      unsorted[z] = original[b]
      b += 1
    end
    z += 1
  end
  while a <= mid
    unsorted[z] = original[a]
    a += 1
    z += 1
  end
  while b <= mid
    unsorted[z] = original[b]
    b += 1
    z += 1
  end
end

def generate_random
  ints = []
  10.times do
    ints << rand(10)
  end
  ints
end

arr = generate_random
p arr
puts "Sorting..."
mergesort(arr, 0, arr.length-1)
p arr

File: Sorting/MergeSort.java
import java.util.*;

public class MergeSort {

    public static void main(String[] args) {

        System.out.println("Generating random array...");
        int[] unsorted = generateRandomArr();
        printArr(unsorted);
        System.out.println("Sorting...");
        mergeSort(unsorted, 0, unsorted.length-1);
        printArr(unsorted);

    } //main

    public static int[] generateRandomArr() {

        int[] random = new int[(int)(Math.random() * 50)];
        for (int i = 0; i < random.length; i++) {
            random[i] = (int)(Math.random() * 100);
        }
        return random;

    } //generateRandomArr

    public static void printArr(int[] arr) {

        System.out.print("[ ");
        for (int i = 0; i < arr.length; i++) {
            System.out.print(arr[i] + " ");
        }
        System.out.println("]");

    } //printArr

    public static void mergeSort(int[] unsorted, int lowIndex, int highIndex) {

        if (lowIndex < highIndex) {
            int mid = lowIndex + (highIndex - lowIndex)/2; //get the index of the center element
            mergeSort(unsorted, lowIndex, mid);
            mergeSort(unsorted, mid + 1, highIndex);
            merge(unsorted, lowIndex, mid, highIndex);
        }

    } // mergeSort

    public static void merge(int[] unsorted, int low, int mid, int  high) {

        int[] original = new int[unsorted.length]; //want a copy of the array that we won't modify
        for (int i = low; i <= high; i++) {
            original[i] = unsorted[i];
        }

        int a = low;
        int b = mid + 1;
        int z = low;

        while (a <= mid && b <= high) {
            if (original[a] <= original[b]) {
                unsorted[z] = original[a];
                a++;
            } else {
                unsorted[z] = original[b];
                b++;
            }
            z++;
        }

        while (a <= mid) {
            unsorted[z] = original[a];
            a++;
            z++;
        }
        while (b <= high) {
            unsorted[z] = original[b];
            b++;
            z++;
        }

    } //merge

} //MergeSort




